File: lib/src/main/java/io/github/kory33/guardedqueries/core/utils/extensions/IteratorExtensions.java
package io.github.kory33.guardedqueries.core.utils.extensions;

import org.apache.commons.lang3.tuple.Pair;

import java.util.Iterator;
import java.util.stream.LongStream;
import java.util.stream.Stream;
import java.util.stream.StreamSupport;

public class IteratorExtensions {
    private IteratorExtensions() {
    }

    /**
     * Zip two iterators together.
     */
    public static <L, R> Iterator<Pair<L, R>> zip(final Iterator<L> leftIterator, final Iterator<R> rightIterator) {
        return new Iterator<>() {
            @Override
            public boolean hasNext() {
                return leftIterator.hasNext() && rightIterator.hasNext();
            }

            @Override
            public Pair<L, R> next() {
                return Pair.of(leftIterator.next(), rightIterator.next());
            }
        };
    }

    /**
     * Zip the iterator together with an infinite Long stream starting from 0.
     */
    public static <T> Iterator<Pair<T, Long>> zipWithIndex(final Iterator<T> iterator) {
        return zip(iterator, LongStream.range(0, Long.MAX_VALUE).iterator());
    }

    /**
     * Convert an iterator to a stream.
     */
    public static <T> Stream<T> stream(final Iterator<T> iterator) {
        final Iterable<T> iterable = () -> iterator;
        return StreamSupport.stream(iterable.spliterator(), false);
    }
}

File: lib/src/main/scala/io/github/kory33/guardedqueries/core/utils/extensions/IteratorExtensions.scala
package io.github.kory33.guardedqueries.core.utils.extensions

import org.apache.commons.lang3.tuple.Pair
import java.util
import java.util.stream.LongStream
import java.util.stream.Stream
import java.util.stream.StreamSupport

object IteratorExtensions {

  /**
   * Convert an iterator to a stream.
   */
  def intoStream[T](iterator: util.Iterator[T]): Stream[T] = {
    val iterable: java.lang.Iterable[T] = () => iterator
    StreamSupport.stream(iterable.spliterator, false)
  }

  def mapInto[R, T](iterator: util.Iterator[_ <: T], mapper: T => R): util.Iterator[R] =
    new util.Iterator[R]() {
      override def hasNext: Boolean = iterator.hasNext

      override def next: R = mapper.apply(iterator.next)
    }
}
